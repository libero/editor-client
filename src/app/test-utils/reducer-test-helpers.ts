import { EditorState } from 'prosemirror-state';
import { Schema } from 'prosemirror-model';

import { Manuscript } from 'app/models/manuscript';
import { getInitialHistory, getLoadableStateSuccess } from 'app/utils/state.utils';
import { ManuscriptHistoryState } from 'app/store';

export function givenState(overrides: Partial<Manuscript>): ManuscriptHistoryState {
  return getLoadableStateSuccess(
    getInitialHistory({
      title: createDummyEditorState(),
      abstract: createDummyEditorState(),
      authors: [],
      affiliations: [],
      keywordGroups: {},
      articleInfo: {
        dtd: '',
        articleType: '',
        articleDOI: ''
      },
      references: [],
      ...overrides
    })
  );
}

export function createDummyEditorState(): EditorState {
  return EditorState.create({
    schema: new Schema({
      nodes: {
        doc: { content: 'inline*' },
        text: { inline: true, group: 'inline' }
      }
    }),
    plugins: []
  });
}
