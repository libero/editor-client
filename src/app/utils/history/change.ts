import { Manuscript } from 'app/types/manuscript';

export abstract class Change {
  abstract applyChange(manuscript: Manuscript): Manuscript;
  abstract rollbackChange(manuscript: Manuscript): Manuscript;
  abstract get isEmpty(): boolean;

  private _timestamp: number;
  constructor() {
    this._timestamp = Date.now();
  }

  get timestamp(): number {
    return this._timestamp;
  }
}

export class BatchChange extends Change {
  private changes: Change[];
  constructor(changes: Change[] = []) {
    super();
    this.changes = changes.filter((change) => !change.isEmpty);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.changes) || this.changes.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduce(
      (acc: Manuscript, change: Change) => {
        return change.applyChange(acc);
      },
      { ...manuscript }
    );
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduceRight(
      (acc: Manuscript, change: Change) => {
        return change.rollbackChange(acc);
      },
      { ...manuscript }
    );
  }
}
