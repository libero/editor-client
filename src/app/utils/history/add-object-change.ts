import { get, set } from 'lodash';

import { Manuscript } from 'app/types/manuscript';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';
import { JSONObject } from 'app/types/utility.types';
import { manuscriptEntityToJson } from 'app/utils/changes.utils';

export class AddObjectChange<T> extends Change {
  constructor(private path: string, private object: T, private idField: string) {
    super();
  }

  get isEmpty(): boolean {
    return false;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to make AddObject change on a non-array section');
    }

    return set(cloneManuscript(manuscript), this.path, [...originalSection, this.object]);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to rollback AddObject change on a non-array section');
    }
    const updatedSection = originalSection.filter((item) => item[this.idField] !== this.object[this.idField]);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  toJSON(): JSONObject {
    return {
      type: 'add-object',
      timestamp: this.timestamp,
      path: this.path,
      idField: this.idField,
      object: manuscriptEntityToJson<T>(this.object)
    };
  }
}
