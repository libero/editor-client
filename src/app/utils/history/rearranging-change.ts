import { get, set } from 'lodash';

import { Manuscript } from '../../types/manuscript';
import { Change } from './change';
import { cloneManuscript } from '../state.utils';
import { JSONObject } from '../../types/utility.types';

export class RearrangingChange extends Change {
  public static createFromListRearrange<T>(path: string, oldOrder: Array<T>, newOrder: Array<T>): RearrangingChange {
    const order = newOrder.map((item) => oldOrder.indexOf(item));
    return new RearrangingChange(path, order);
  }

  public static createFromItemMoved<T>(
    path: string,
    oldIndex: number,
    newIndex: number,
    collection: Array<T>
  ): RearrangingChange {
    const order = new Array(collection.length).fill(undefined).map((_, index) => index);
    order.splice(oldIndex, 1);
    order.splice(newIndex, 0, oldIndex);
    return new RearrangingChange(path, order);
  }

  public static fromJSON<T>(data: JSONObject): RearrangingChange {
    const change = new RearrangingChange(data.path as string, data.order as Array<number>);
    change._timestamp = data.timestamp as number;
    return change;
  }

  private constructor(private path: string, private order: number[]) {
    super();
  }

  get isEmpty(): boolean {
    return !this.order.some((el, index) => el !== index);
  }

  isPathAffected(pathPattern: RegExp): boolean {
    return pathPattern.test(this.path);
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return set(cloneManuscript(manuscript), this.path, this.applyOrder(get(manuscript, this.path), this.order));
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const rollbackOrder = this.reverseOrder(this.order);
    return set({ ...manuscript }, this.path, this.applyOrder(get(manuscript, this.path), rollbackOrder));
  }

  toJSON(): JSONObject {
    return {
      type: 'rearranging',
      timestamp: this.timestamp,
      path: this.path,
      order: this.order
    };
  }

  private applyOrder<T>(collection: Array<T>, order: number[]): Array<T> {
    return order.map((index) => collection[index]);
  }

  private reverseOrder(order: number[]): number[] {
    return order.reduce((acc, oldIndex, newIndex) => {
      acc[oldIndex] = newIndex;
      return acc;
    }, new Array<number>(order.length));
  }
}
