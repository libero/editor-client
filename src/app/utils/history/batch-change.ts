import { Manuscript } from 'app/types/manuscript';
import { JSONObject } from 'app/types/utility.types';
import { deserializeChanges } from 'app/utils/changes.utils';
import { Change } from 'app/utils/history/change';

export class BatchChange extends Change {
  static fromJSON(data: JSONObject): BatchChange {
    const change = new BatchChange(deserializeChanges(data.changes as JSONObject[]));
    change._timestamp = data.timestamp as number;
    return change;
  }

  private changes: Change[];
  constructor(changes: Change[] = []) {
    super();
    this.changes = changes.filter((change) => !change.isEmpty);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.changes) || this.changes.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduce((acc: Manuscript, change: Change) => {
      return change.applyChange(acc);
    }, manuscript);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    return this.changes.reduceRight((acc: Manuscript, change: Change) => {
      return change.rollbackChange(acc);
    }, manuscript);
  }

  toJSON(): JSONObject {
    return {
      type: 'batch',
      changes: this.changes.map((change) => change.toJSON()),
      timestamp: this.timestamp
    };
  }
}
