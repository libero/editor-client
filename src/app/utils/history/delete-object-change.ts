import { Manuscript } from 'app/types/manuscript';
import { get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class DeleteObjectChange<T> implements Change {
  private removedIndex: number;
  constructor(private path: string, private object: T, private idField: string) {}

  get isEmpty(): boolean {
    return false;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to make AddObject change on a non-array section');
    }
    this.removedIndex = originalSection.findIndex((item) => item[this.idField] === this.object[this.idField]);
    const updatedSection = [...originalSection];
    updatedSection.splice(this.removedIndex, 1);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    if (!Array.isArray(originalSection)) {
      throw new TypeError('Trying to rollback AddObject change on a non-array section');
    }
    const updatedSection = originalSection.filter((item) => item[this.idField] === this.object[this.idField]);
    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }
}
