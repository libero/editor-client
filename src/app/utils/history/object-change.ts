import * as deepDiff from 'deep-diff';
import { Manuscript } from 'app/types/manuscript';
import { cloneDeep, get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class ObjectChange<T> implements Change {
  private differences: deepDiff.Diff<T, T>[];

  constructor(private path: string, oldObject: T, newObject: T) {
    this.differences = deepDiff.diff(oldObject, newObject);
  }

  get isEmpty(): boolean {
    return !Array.isArray(this.differences) || this.differences.length === 0;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = cloneDeep(acc);
      deepDiff.applyChange(newAcc, acc, diff);
      return newAcc;
    }, cloneDeep(originalSection));

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const originalSection = get(manuscript, this.path);

    const updatedSection = this.differences.reduce((acc: T, diff) => {
      const newAcc = cloneDeep(acc);
      deepDiff.revertChange(newAcc, acc, diff);
      return newAcc;
    }, originalSection);

    return set(cloneManuscript(manuscript), this.path, updatedSection);
  }
}
