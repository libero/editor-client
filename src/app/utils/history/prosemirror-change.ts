import { Transaction } from 'prosemirror-state';
import { Manuscript } from 'app/types/manuscript';
import { get, set } from 'lodash';
import { Change } from 'app/utils/history/change';
import { cloneManuscript } from 'app/utils/state.utils';

export class ProsemirrorChange implements Change {
  constructor(private path: string, private transaction: Transaction) {}

  get isEmpty(): boolean {
    return !this.transaction.docChanged;
  }

  applyChange(manuscript: Manuscript): Manuscript {
    const editorState = get(manuscript, this.path);
    return set(cloneManuscript(manuscript), this.path, editorState.apply(this.transaction));
  }

  rollbackChange(manuscript: Manuscript): Manuscript {
    const invertedSteps = this.transaction.steps.map((step, index) => {
      return step.invert(this.transaction.docs[index]);
    });

    const editorState = get(manuscript, this.path);
    const rollbackTransaction = editorState.tr;
    invertedSteps.reverse().forEach((step) => rollbackTransaction.maybeStep(step));

    return set(cloneManuscript(manuscript), this.path, editorState.apply(rollbackTransaction));
  }
}
