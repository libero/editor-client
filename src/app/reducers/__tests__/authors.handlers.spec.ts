import { cloneDeep } from 'lodash';

import { givenState } from 'app/test-utils/reducer-test-helpers';
import { addAuthor, deleteAuthor, moveAuthor, updateAuthor } from 'app/reducers/authors.handlers';
import { BatchChange } from 'app/utils/history/change';

describe('Authors reducers', () => {
  it('should add new author', () => {
    const state = givenState({});
    const newAuthor = { id: 'newId', firstName: 'Jules', lastName: 'Verne', affiliations: [] };
    const updatedState = cloneDeep(state);
    updatedState.data.present.authors = [newAuthor];
    updatedState.data.past = [expect.any(BatchChange)];

    const newState = addAuthor(state, newAuthor);
    expect(newState).toEqual(updatedState);
  });

  it('should update existing author', () => {
    const state = givenState({
      authors: [{ id: 'newId', firstName: 'Jules', lastName: 'Verne', affiliations: [] }]
    });
    const updatedAuthor = { id: 'newId', firstName: 'Jules Gabriel', lastName: 'Verne', affiliations: [] };
    const updatedState = cloneDeep(state);
    updatedState.data.present.authors = [updatedAuthor];
    updatedState.data.past = [expect.any(BatchChange)];

    const newState = updateAuthor(state, updatedAuthor);
    expect(newState).toEqual(updatedState);
  });

  it('should move author', () => {
    const state = givenState({
      authors: [
        { id: 'id1', firstName: 'Jules', lastName: 'Verne', affiliations: [] },
        { id: 'id2', firstName: 'H G', lastName: 'Wells', affiliations: [] }
      ]
    });

    const updatedState = cloneDeep(state);
    updatedState.data.present.authors.reverse();
    updatedState.data.past = [expect.any(BatchChange)];
    const newState = moveAuthor(state, { index: 1, author: state.data.present.authors[0] });
    expect(newState).toEqual(updatedState);
  });

  it('should delete author', () => {
    const state = givenState({
      authors: [
        { id: 'id1', firstName: 'Jules', lastName: 'Verne', affiliations: [] },
        { id: 'id2', firstName: 'H G', lastName: 'Wells', affiliations: [] }
      ]
    });

    const updatedState = cloneDeep(state);
    updatedState.data.present.authors = updatedState.data.present.authors.slice(1);
    updatedState.data.past = [expect.any(BatchChange)];
    const newState = deleteAuthor(state, state.data.present.authors[0]);
    expect(newState).toEqual(updatedState);
  });
});
