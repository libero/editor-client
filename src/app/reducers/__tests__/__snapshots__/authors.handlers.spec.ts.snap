// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Authors reducers should add new author 1`] = `
Object {
  "data": Object {
    "future": Array [],
    "past": Array [
      Object {
        "changes": Array [
          Object {
            "idField": "id",
            "object": Person {
              "_id": "newId",
              "affiliations": Array [],
              "bio": Object {
                "doc": Object {
                  "content": Array [
                    Object {
                      "type": "paragraph",
                    },
                  ],
                  "type": "doc",
                },
                "selection": Object {
                  "anchor": 1,
                  "head": 1,
                  "type": "text",
                },
              },
              "competingInterestStatement": undefined,
              "email": undefined,
              "firstName": "Jules",
              "hasCompetingInterest": undefined,
              "isAuthenticated": undefined,
              "isCorrespondingAuthor": undefined,
              "lastName": "Verne",
              "orcid": undefined,
              "suffix": undefined,
            },
            "path": "authors",
            "timestamp": 1487076708000,
            "type": "add-object",
          },
          Object {
            "changes": Array [
              Object {
                "differences": Array [
                  DiffEdit {
                    "kind": "E",
                    "lhs": undefined,
                    "path": Array [
                      "copyrightStatement",
                    ],
                    "rhs": "",
                  },
                ],
                "path": "articleInfo",
                "timestamp": 1487076708000,
                "type": "update-object",
              },
            ],
            "timestamp": 1487076708000,
            "type": "batch",
          },
        ],
        "timestamp": 1487076708000,
        "type": "batch",
      },
    ],
    "present": Object {
      "abstract": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "acknowledgements": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "affiliations": Array [],
      "articleInfo": ArticleInformation {
        "_id": "some_uuid",
        "articleDOI": "",
        "articleType": "",
        "copyrightStatement": "",
        "dtd": "",
        "elocationId": "",
        "licenseText": Object {
          "doc": Object {
            "type": "doc",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
        "licenseType": "CC0",
        "publicationDate": "",
        "publisherId": "",
        "subjects": Array [],
        "volume": "",
      },
      "authors": Array [
        Person {
          "_id": "newId",
          "affiliations": Array [],
          "bio": Object {
            "doc": Object {
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "doc",
            },
            "selection": Object {
              "anchor": 1,
              "head": 1,
              "type": "text",
            },
          },
          "competingInterestStatement": undefined,
          "email": undefined,
          "firstName": "Jules",
          "hasCompetingInterest": undefined,
          "isAuthenticated": undefined,
          "isCorrespondingAuthor": undefined,
          "lastName": "Verne",
          "orcid": undefined,
          "suffix": undefined,
        },
      ],
      "body": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "impactStatement": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "journalMeta": Object {
        "issn": "",
        "publisherName": "",
      },
      "keywordGroups": Object {},
      "references": Array [],
      "relatedArticles": Array [],
      "title": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
    },
  },
  "error": undefined,
  "isLoading": false,
}
`;

exports[`Authors reducers should delete author 1`] = `
Object {
  "data": Object {
    "future": Array [],
    "past": Array [
      Object {
        "changes": Array [
          Object {
            "idField": "id",
            "object": Person {
              "_id": "id1",
              "affiliations": Array [],
              "bio": Object {
                "doc": Object {
                  "content": Array [
                    Object {
                      "type": "paragraph",
                    },
                  ],
                  "type": "doc",
                },
                "selection": Object {
                  "anchor": 1,
                  "head": 1,
                  "type": "text",
                },
              },
              "competingInterestStatement": undefined,
              "email": undefined,
              "firstName": "Jules",
              "hasCompetingInterest": undefined,
              "isAuthenticated": undefined,
              "isCorrespondingAuthor": undefined,
              "lastName": "Verne",
              "orcid": undefined,
              "suffix": undefined,
            },
            "path": "authors",
            "removedIndex": 0,
            "timestamp": 1487076708000,
            "type": "delete-object",
          },
          Object {
            "changes": Array [
              Object {
                "differences": Array [
                  DiffEdit {
                    "kind": "E",
                    "lhs": undefined,
                    "path": Array [
                      "copyrightStatement",
                    ],
                    "rhs": "",
                  },
                ],
                "path": "articleInfo",
                "timestamp": 1487076708000,
                "type": "update-object",
              },
            ],
            "timestamp": 1487076708000,
            "type": "batch",
          },
        ],
        "timestamp": 1487076708000,
        "type": "batch",
      },
    ],
    "present": Object {
      "abstract": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "acknowledgements": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "affiliations": Array [],
      "articleInfo": ArticleInformation {
        "_id": "some_uuid",
        "articleDOI": "",
        "articleType": "",
        "copyrightStatement": "",
        "dtd": "",
        "elocationId": "",
        "licenseText": Object {
          "doc": Object {
            "type": "doc",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
        "licenseType": "CC0",
        "publicationDate": "",
        "publisherId": "",
        "subjects": Array [],
        "volume": "",
      },
      "authors": Array [
        Person {
          "_id": "id2",
          "affiliations": Array [],
          "bio": Object {
            "doc": Object {
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "doc",
            },
            "selection": Object {
              "anchor": 1,
              "head": 1,
              "type": "text",
            },
          },
          "competingInterestStatement": undefined,
          "email": undefined,
          "firstName": "H G",
          "hasCompetingInterest": undefined,
          "isAuthenticated": undefined,
          "isCorrespondingAuthor": undefined,
          "lastName": "Wells",
          "orcid": undefined,
          "suffix": undefined,
        },
      ],
      "body": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "impactStatement": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "journalMeta": Object {
        "issn": "",
        "publisherName": "",
      },
      "keywordGroups": Object {},
      "references": Array [],
      "relatedArticles": Array [],
      "title": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
    },
  },
  "error": undefined,
  "isLoading": false,
}
`;

exports[`Authors reducers should move author 1`] = `
Object {
  "data": Object {
    "future": Array [],
    "past": Array [
      Object {
        "changes": Array [
          Object {
            "order": Array [
              1,
              0,
            ],
            "path": "authors",
            "timestamp": 1487076708000,
            "type": "rearranging",
          },
          Object {
            "changes": Array [
              Object {
                "differences": Array [
                  DiffEdit {
                    "kind": "E",
                    "lhs": undefined,
                    "path": Array [
                      "copyrightStatement",
                    ],
                    "rhs": "",
                  },
                ],
                "path": "articleInfo",
                "timestamp": 1487076708000,
                "type": "update-object",
              },
            ],
            "timestamp": 1487076708000,
            "type": "batch",
          },
        ],
        "timestamp": 1487076708000,
        "type": "batch",
      },
    ],
    "present": Object {
      "abstract": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "acknowledgements": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "affiliations": Array [],
      "articleInfo": ArticleInformation {
        "_id": "some_uuid",
        "articleDOI": "",
        "articleType": "",
        "copyrightStatement": "",
        "dtd": "",
        "elocationId": "",
        "licenseText": Object {
          "doc": Object {
            "type": "doc",
          },
          "selection": Object {
            "anchor": 0,
            "head": 0,
            "type": "text",
          },
        },
        "licenseType": "CC0",
        "publicationDate": "",
        "publisherId": "",
        "subjects": Array [],
        "volume": "",
      },
      "authors": Array [
        Person {
          "_id": "id2",
          "affiliations": Array [],
          "bio": Object {
            "doc": Object {
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "doc",
            },
            "selection": Object {
              "anchor": 1,
              "head": 1,
              "type": "text",
            },
          },
          "competingInterestStatement": undefined,
          "email": undefined,
          "firstName": "H G",
          "hasCompetingInterest": undefined,
          "isAuthenticated": undefined,
          "isCorrespondingAuthor": undefined,
          "lastName": "Wells",
          "orcid": undefined,
          "suffix": undefined,
        },
        Person {
          "_id": "id1",
          "affiliations": Array [],
          "bio": Object {
            "doc": Object {
              "content": Array [
                Object {
                  "type": "paragraph",
                },
              ],
              "type": "doc",
            },
            "selection": Object {
              "anchor": 1,
              "head": 1,
              "type": "text",
            },
          },
          "competingInterestStatement": undefined,
          "email": undefined,
          "firstName": "Jules",
          "hasCompetingInterest": undefined,
          "isAuthenticated": undefined,
          "isCorrespondingAuthor": undefined,
          "lastName": "Verne",
          "orcid": undefined,
          "suffix": undefined,
        },
      ],
      "body": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "impactStatement": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
      "journalMeta": Object {
        "issn": "",
        "publisherName": "",
      },
      "keywordGroups": Object {},
      "references": Array [],
      "relatedArticles": Array [],
      "title": Object {
        "doc": Object {
          "type": "doc",
        },
        "selection": Object {
          "anchor": 0,
          "head": 0,
          "type": "text",
        },
      },
    },
  },
  "error": undefined,
  "isLoading": false,
}
`;
